ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel/Genesis.hs

```diff
@@ -0,0 +1,263 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | An implementation of the prefix selection algorithm.
+module Ouroboros.Consensus.NodeKernel.Genesis (
+    processWithPrefixSelection
+    -- * For testing purposes
+  , prefixSelection
+  ) where
+
+import           Prelude hiding (length)
+
+import           Data.Function (on)
+import           Data.List (foldl', maximumBy)
+import           Data.Map (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (catMaybes)
+
+import           Cardano.Slotting.Slot
+
+import           Ouroboros.Consensus.Block.Abstract (Header, blockSlot)
+import           Ouroboros.Consensus.Block.SupportsProtocol
+                     (BlockSupportsProtocol)
+
+import           Ouroboros.Consensus.Config.GenesisWindowLength
+import           Ouroboros.Network.AnchoredFragment hiding (filter)
+import           Ouroboros.Network.AnchoredFragment.Completeness
+import           Ouroboros.Network.AnchoredSeq hiding (filter, join, map)
+
+{-------------------------------------------------------------------------------
+ The prefix selection algorithm
+-------------------------------------------------------------------------------}
+
+-- | Process a map of candidates as stated in the prefixSelection chapter of the
+-- report.
```

Comment:
I wonder if we want to simply refer to the report. I can imagine a reader might find it inconvenient to have to context switch to a different document and find where the relevant information is.